File: CPP/No692.cc
/**
 * Created by Xiaozhong on 2020/11/18.
 * Copyright (c) 2020/11/18 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<string> topKFrequent(vector<string> &words, int k) {
        map<string, int> mCounter;
        for (const auto word : words) mCounter[word]++;
        vector<pair<string, int>> vCounter;
        for (auto iter = mCounter.begin(); iter != mCounter.end(); ++iter) {
            vCounter.emplace_back(make_pair(iter->first, iter->second));
        }
        sort(vCounter.begin(), vCounter.end(), [](const pair<string, int> &rhs, const pair<string, int> &lhs) {
            return rhs.second == lhs.second ? rhs.first < lhs.first : rhs.second > lhs.second;
        });
        vector<string> ans;
        for (int i = 0; i < k; ++i) {
            ans.emplace_back(vCounter[i].first);
        }
        return ans;
    }
};

File: Java/src/No692.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.*;

public class No692 {
    public List<String> topKFrequent(String[] words, int k) {
        Map<String, Integer> counter = new HashMap<>();
        for (String word : words) {
            counter.put(word, counter.getOrDefault(word, 0) + 1);
        }
        PriorityQueue<Map.Entry<String, Integer>> queue = new PriorityQueue<>((o1, o2) -> o2.getValue() - o1.getValue() == 0 ? o1.getKey().compareTo(o2.getKey()) : o2.getValue() - o1.getValue());
        for (Map.Entry<String, Integer> entry : counter.entrySet()) queue.offer(entry);
        List<String> ans = new ArrayList<>();
        for (int i = 0; i < k; ++i) ans.add(queue.poll().getKey());
        return ans;
    }
}
